File: modules/expo-image-resize/android/src/main/java/expo/modules/imageresize/ExpoImageResizeModule.kt
package expo.modules.imageresize

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.ColorSpace
import android.graphics.ImageDecoder
import android.net.Uri
import android.os.Build
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File
import java.io.FileOutputStream

/**
 * Resize a JPEG file on disk to a target width, preserving aspect ratio,
 * and write the result as a JPEG at the requested quality.
 *
 * Decode strategy (Phases 3 + 4):
 *   1. Fast path: two-pass BitmapFactory (measure + sampled decode).
 *      Covers the vast majority of well-formed JPEG / PNG / WebP
 *      sources at low memory cost.
 *   2. Fallback path: ImageDecoder.decodeBitmap (API 28+). Used when
 *      BitmapFactory returns null — typically CMYK JPEGs or PNGs
 *      with unusual ICC profiles that BitmapFactory's JNI bridge
 *      can't decode. ImageDecoder has broader format coverage and
 *      lets us request explicit ARGB_8888 + sRGB.
 *   3. Always-sanitise pre-flight: every decoded bitmap is forced
 *      to ARGB_8888 (32-bit RGBA), and on API 26+ to the sRGB
 *      colour space, so the JPEG we write is in a colour space
 *      every downstream consumer handles cleanly.
 *
 * Deliberately doesn't use Glide / coroutines — `expo-image-manipulator`
 * Glide-backed loader has a double-resume race that surfaces as
 * `IllegalStateException: alreadyResumed` on Android 16 / tight-
 * lifecycle ROMs. `AsyncFunction` dispatches each call to the Expo
 * module background thread automatically.
 */
class ExpoImageResizeModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoImageResize")

    AsyncFunction("resizeImageToFileAsync") { sourceUri: String, targetUri: String, maxWidth: Int, quality: Double ->
      val sourcePath = resolvePath(sourceUri)
        ?: throw Exception("Invalid source URI: $sourceUri")
      val targetPath = resolvePath(targetUri)
        ?: throw Exception("Invalid target URI: $targetUri")

      val sourceFile = File(sourcePath)
      val decoded = decodeSource(sourceFile, maxWidth)
        ?: throw Exception("Failed to decode bitmap: $sourcePath")

      // Scale to exact target width; height preserves aspect ratio.
      val targetWidth = maxWidth.coerceAtLeast(1)
      val aspect = decoded.height.toFloat() / decoded.width.toFloat()
      val targetHeight = (targetWidth * aspect).toInt().coerceAtLeast(1)
      val scaled = if (decoded.width == targetWidth && decoded.height == targetHeight) {
        decoded
      } else {
        Bitmap.createScaledBitmap(decoded, targetWidth, targetHeight, true)
      }

      // Ensure parent directory exists before writing.
      val targetFile = File(targetPath)
      targetFile.parentFile?.mkdirs()

      try {
        FileOutputStream(targetFile).use { out ->
          val q = (quality.coerceIn(0.0, 1.0) * 100).toInt()
          scaled.compress(Bitmap.CompressFormat.JPEG, q, out)
        }
      } finally {
        if (scaled !== decoded) scaled.recycle()
        decoded.recycle()
      }
    }
  }

  /**
   * Two-phase decode + sanitise pipeline.
   *
   * Returns a non-null `Bitmap` in ARGB_8888 colour config (and sRGB
   * colour space on API 26+) or null if every available decoder
   * refused the source.
   */
  private fun decodeSource(file: File, maxWidth: Int): Bitmap? {
    // Phase 1 fast path: BitmapFactory with inSampleSize so we never
    // hold the full-resolution bitmap when maxWidth is much smaller
    // than the source.
    val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
    BitmapFactory.decodeFile(file.absolutePath, bounds)
    if (bounds.outWidth > 0 && bounds.outHeight > 0) {
      val sampleSize = calculateInSampleSize(bounds.outWidth, maxWidth)
      val decodeOpts = BitmapFactory.Options().apply {
        inSampleSize = sampleSize
        inPreferredConfig = Bitmap.Config.ARGB_8888
      }
      val fast = BitmapFactory.decodeFile(file.absolutePath, decodeOpts)
      if (fast != null) {
        return normaliseColorSpace(fast)
      }
    }

    // Phase 3 fallback: ImageDecoder is available on API 28+. Broader
    // format coverage than BitmapFactory — handles CMYK JPEGs and
    // ICC-profile-heavy PNGs that the older API rejects. We force
    // ARGB_8888 + (on API 26+) sRGB during the listener callback so
    // any colour-space conversion happens inside the decoder.
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
      val source = ImageDecoder.createSource(file)
      try {
        val decoded = ImageDecoder.decodeBitmap(source) { decoder, info, _ ->
          decoder.allocator = ImageDecoder.ALLOCATOR_SOFTWARE
          // Constrain output dimensions so we don't hold a huge
          // bitmap when maxWidth is small.
          val srcW = info.size.width
          val srcH = info.size.height
          if (srcW > 0 && srcH > 0 && maxWidth > 0 && srcW > maxWidth) {
            val aspect = srcH.toFloat() / srcW.toFloat()
            decoder.setTargetSize(maxWidth, (maxWidth * aspect).toInt().coerceAtLeast(1))
          }
          if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            decoder.setTargetColorSpace(ColorSpace.get(ColorSpace.Named.SRGB))
          }
        }
        return normaliseColorSpace(decoded)
      } catch (e: Exception) {
        // ImageDecoder rejected too — let the caller throw.
      }
    }

    return null
  }

  /**
   * Phase 4 belt-and-braces: ensure the bitmap we return is in
   * ARGB_8888 + sRGB regardless of which decode path produced it.
   * `BitmapFactory.inPreferredConfig = ARGB_8888` is a HINT — the
   * decoder may ignore it for certain sources — so we explicitly
   * re-create the bitmap if the actual config / colour space drifts.
   */
  private fun normaliseColorSpace(bitmap: Bitmap): Bitmap {
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
      // Pre-26 has no per-bitmap colour space; ARGB_8888 is the best we can do.
      return if (bitmap.config == Bitmap.Config.ARGB_8888) {
        bitmap
      } else {
        val copy = bitmap.copy(Bitmap.Config.ARGB_8888, false)
        bitmap.recycle()
        copy
      }
    }

    val needsConfigChange = bitmap.config != Bitmap.Config.ARGB_8888
    val srgb = ColorSpace.get(ColorSpace.Named.SRGB)
    val needsColorSpaceChange = bitmap.colorSpace != srgb
    if (!needsConfigChange && !needsColorSpaceChange) {
      return bitmap
    }

    // Bitmap.copy doesn't take a colour space, so build the new bitmap
    // explicitly and draw the original in — that path applies any
    // necessary colour-space conversion.
    val target = Bitmap.createBitmap(
      bitmap.width,
      bitmap.height,
      Bitmap.Config.ARGB_8888,
      bitmap.hasAlpha(),
      srgb,
    )
    val canvas = android.graphics.Canvas(target)
    canvas.drawBitmap(bitmap, 0f, 0f, null)
    bitmap.recycle()
    return target
  }

  private fun resolvePath(uri: String): String? {
    if (uri.startsWith("file://")) {
      return Uri.parse(uri).path
    }
    return uri
  }

  /**
   * Compute a sane inSampleSize for the target width. Android's
   * BitmapFactory only honours powers of 2; `outWidth / maxWidth` rounded
   * down to the nearest power of 2 gives the largest safe step-down
   * without going below the target.
   */
  private fun calculateInSampleSize(sourceWidth: Int, maxWidth: Int): Int {
    if (sourceWidth <= maxWidth || maxWidth <= 0) return 1
    var sample = 1
    while ((sourceWidth / (sample * 2)) >= maxWidth) {
      sample *= 2
    }
    return sample
  }
}

File: modules/expo-image-resize/ios/ExpoImageResizeModule.swift
import ExpoModulesCore
import Foundation
import UIKit
import ImageIO
import CoreGraphics

public class ExpoImageResizeModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoImageResize")

    // Resize a local image to `maxWidth` pixels wide (aspect-preserving)
    // and write the result as a JPEG at `quality` to `targetUri`.
    //
    // Decode strategy (Phases 2 + 4):
    //   1. Fast path: UIImage(contentsOfFile:) — handles JPEG / PNG /
    //      WebP / HEIC etc. that UIImage natively accepts.
    //   2. Fallback path: CGImageSource decode — used when UIImage
    //      returns nil. CGImageSource has broader format coverage and
    //      gives us the raw CGImage we can re-draw through a known
    //      color space, fixing common "decode succeeds but renderer
    //      barfs" issues with CMYK JPEGs, embedded ICC profiles, and
    //      unusual PNG variants.
    //   3. Always-sanitise pre-flight: regardless of which decode path
    //      produced the image, we draw it into a renderer that's pinned
    //      to standard-range sRGB output, so the JPEG we write is in a
    //      colour space every downstream consumer (RN's <Image>, the
    //      OS preview, share sheet, etc.) handles cleanly.
    //
    // `AsyncFunction` dispatches each call to a background queue
    // automatically.
    AsyncFunction("resizeImageToFileAsync") { (sourceUri: String, targetUri: String, maxWidth: Int, quality: Double) in
      let sourcePath = Self.resolvePath(sourceUri)
      let targetPath = Self.resolvePath(targetUri)

      let source = try Self.decodeSource(sourcePath)
      guard source.size.width > 0, source.size.height > 0 else {
        throw ResizeError.invalidDimensions
      }

      let targetWidth = CGFloat(max(1, maxWidth))
      let aspect = source.size.height / source.size.width
      let targetSize = CGSize(
        width: targetWidth,
        height: max(1, (targetWidth * aspect).rounded())
      )

      // Phase 4: pin the renderer to sRGB / standard range so the
      // output JPEG is in a colour space every downstream consumer
      // handles. Without this the renderer follows the device display
      // (which can be P3 on iPhone 7+), producing wider-gamut JPEGs
      // that subtly mis-render under some Android decoders that
      // ingest covers via share/backup-restore flows.
      let format = UIGraphicsImageRendererFormat()
      format.scale = 1
      format.opaque = true
      format.preferredRange = .standard
      let renderer = UIGraphicsImageRenderer(size: targetSize, format: format)
      let resized = renderer.image { _ in
        source.draw(in: CGRect(origin: .zero, size: targetSize))
      }

      let clampedQuality = CGFloat(max(0.0, min(1.0, quality)))
      guard let data = resized.jpegData(compressionQuality: clampedQuality) else {
        throw ResizeError.encodeFailed
      }

      // Ensure parent directory exists before writing.
      let targetUrl = URL(fileURLWithPath: targetPath)
      try? FileManager.default.createDirectory(
        at: targetUrl.deletingLastPathComponent(),
        withIntermediateDirectories: true
      )
      try data.write(to: targetUrl)
    }
  }

  /// Two-phase decode: try UIImage first (fast path covers the vast
  /// majority of cover art), then fall back to a CGImageSource +
  /// sRGB-context redraw for sources UIImage refuses. The fallback
  /// path is the recovery for CMYK JPEGs and PNGs with unusual ICC
  /// profiles where UIImage returns nil rather than the canonical
  /// in-memory representation.
  private static func decodeSource(_ sourcePath: String) throws -> UIImage {
    if let fast = UIImage(contentsOfFile: sourcePath) {
      return fast
    }

    // UIImage refused. Try the lower-level ImageIO decoder.
    let url = URL(fileURLWithPath: sourcePath)
    guard let cgSource = CGImageSourceCreateWithURL(url as CFURL, nil) else {
      throw ResizeError.decodeFailed(sourcePath)
    }
    // Force the source to fully decode immediately, into a known sRGB
    // colour space — both options together cover CMYK conversion and
    // strip embedded profiles that would otherwise carry through.
    let options: [CFString: Any] = [
      kCGImageSourceShouldCache: true,
      kCGImageSourceShouldAllowFloat: false,
    ]
    guard let cgImage = CGImageSourceCreateImageAtIndex(cgSource, 0, options as CFDictionary) else {
      throw ResizeError.decodeFailed(sourcePath)
    }

    // Re-draw into an explicit sRGB CGContext to normalise the colour
    // space — CMYK / Lab / CalRGB sources all collapse to sRGB here.
    let width = cgImage.width
    let height = cgImage.height
    guard width > 0, height > 0 else {
      throw ResizeError.invalidDimensions
    }
    guard let colorSpace = CGColorSpace(name: CGColorSpace.sRGB) else {
      throw ResizeError.decodeFailed(sourcePath)
    }
    let bitmapInfo = CGBitmapInfo(rawValue:
      CGImageAlphaInfo.noneSkipLast.rawValue
    )
    guard let context = CGContext(
      data: nil,
      width: width,
      height: height,
      bitsPerComponent: 8,
      bytesPerRow: 0,
      space: colorSpace,
      bitmapInfo: bitmapInfo.rawValue
    ) else {
      throw ResizeError.decodeFailed(sourcePath)
    }
    context.draw(cgImage, in: CGRect(x: 0, y: 0, width: width, height: height))
    guard let normalised = context.makeImage() else {
      throw ResizeError.decodeFailed(sourcePath)
    }
    return UIImage(cgImage: normalised)
  }

  private static func resolvePath(_ uri: String) -> String {
    if uri.hasPrefix("file://") {
      return URL(string: uri)?.path ?? String(uri.dropFirst("file://".count))
    }
    return uri
  }
}

private enum ResizeError: Error, LocalizedError {
  case decodeFailed(String)
  case encodeFailed
  case invalidDimensions

  var errorDescription: String? {
    switch self {
    case .decodeFailed(let path): return "Failed to decode image at \(path)"
    case .encodeFailed: return "Failed to encode JPEG"
    case .invalidDimensions: return "Source image has zero dimensions"
    }
  }
}
